feat(defaults): store spellfile in stdpath('data')

yochem opened the pull request:
Problem:
Default spellfile location rtp[0]/spell/ is inconsistent and does not conform to the XDG base spec.

Solution:
Default to `stdpath("data") .. "/spell/"` if 'spellfile' is unset.

Closes: #32958.

yochem committed:
feat(spellfile)!: Default to stdpath('data') for spell additions

Problem:
Defaulting to rtp[0]/spell ends up with inconsistent spell addition
files. `stdpath('config')/spell` (generally the first rtp path) also
does not conform to the XDG base spec.

Solution:
Use `stdpath('data')/spell/{spelllang[0]}.{encoding}.add` as default
'spellfile' location if none is given.

justinmk commented on src/nvim/spellfile.c:
```diff
@@ -5569,48 +5570,39 @@ static void init_spellfile(void)
   }
+  
+  // If the spelllang points to a .spl file, use that path directly.
+  if (aspath) {
+    xmemcpyz(buf, curbuf->b_s.b_p_spl, (size_t)(lstart - curbuf->b_s.b_p_spl - 1));
+  } else {
+    buf = get_xdg_home(kXDGDataHome);
+  }
```
would need to free `buf` before reassigning it here. But instead of doing that, change this to:

```suggestion
  char *buf = aspath? xmalloc(MAXPATHL) : get_xdg_home(kXDGDataHome);
```

yochem committed:
feat(spellfile): stdpath('data') as default location

Problem:
First rtp directory is unpredictable and not in line with XDG
base spec.

Solution:
Use stdpath('data')/spell as directory if 'spellfile' is not set.

yochem committed:
Remove unnecessary include

justinmk commented on test/functional/core/spellfile_spec.lua:
```diff
@@ -11,15 +11,18 @@ local write_file = t.write_file
 describe('spellfile', function()
   before_each(function()
     clear()
+    fn.setenv('XDG_DATA_HOME', old_xdg_data)
```
instead of this, just call clear() with `env`:

    clear{ env = { ... } }

then you don't need to save/restore `old_xdg_data`.

yochem replied:
Perfect, thank you!

yochem committed:
Reset ENV during test with clear()

justinmk commented on runtime/doc/news.txt:
```diff
@@ -172,7 +172,8 @@ CHANGED FEATURES                                                 *news-changed*
 These existing features changed their behavior.
 
-• todo
+• |spellfile| uses `stdpath('data')/spell` as default location instead of the
```
```suggestion
• |spellfile| location defaults to `stdpath('data').."/spell"` instead of the
```

justinmk commented on src/nvim/spellfile.c:
```diff
@@ -5570,49 +5568,45 @@ static void init_spellfile(void)
-        if (os_file_is_writable(buf) != 2) {
-          os_mkdir(buf, 0755);
-        }
+  char *buf = xmalloc(MAXPATHL);
```
asan

```
= ==13704==ERROR: LeakSanitizer: detected memory leaks
= 
= Direct leak of 4096 byte(s) in 1 object(s) allocated from:
=     #0 0x564ea818dd93 in malloc (/home/runner/work/neovim/neovim/build/bin/nvim+0xd03d93) (BuildId: 839761a9c2c58f0050a91741c7578dac1ff5726f)
=     #1 0x564ea8de5e24 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:98:15
=     #2 0x564ea8de6014 in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:132:15
=     #3 0x564ea9385cf8 in init_spellfile /home/runner/work/neovim/neovim/src/nvim/spellfile.c:5571:15
=     #4 0x564ea9383d39 in spell_add_word /home/runner/work/neovim/neovim/src/nvim/spellfile.c:5419:7
```

justinmk commented on test/functional/core/spellfile_spec.lua:
```diff
@@ -117,4 +117,27 @@ describe('spellfile', function()
     api.nvim_set_option_value('spellfile', fname, {})
   end)
+
+  describe('default', function ()
```
```suggestion
  describe('default location', function ()
```

justinmk commented on runtime/doc/options.txt:
```diff
@@ -5755,11 +5755,10 @@ A jump table for the options with a short description can be found at |Q_op|.
-	there is no "spell" directory yet it will be created.  For the file
-	name the first language name that appears in 'spelllang' is used,
-	ignoring the region.
+	When a word is added while this option is empty Nvim will use
+	`stdpath('data')/spell`.  If there is no "spell" directory yet it will
```
```suggestion
	When a word is added while this option is empty Nvim will use
	(and auto-create) `stdpath('data')/spell/`. For the file name ...
```

justinmk reviewed (approved):
lgtm after asan is fixed

yochem committed:
Update docs text according to reviews

yochem commented:
Thanks for the review. I've resolved all doc remarks. I'm currently looking into the asan error, but can't get asan to work locally? I've tried the examples in the [contributing docs](https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#sanitizers-asan-and-ubsan), but /tmp/nvim.{pid} or /tmp/nvim_asan don't apppear (nvim also doesn't crash). How can I check if my current solution fixes the memory leak? 

I also noticed an old test failing because it tests on the old Vim behaviour. Should I disable / skip / delete this test? It's `Test_init_spellfile` from test/old/testdir/test_spellfile.vim.

yochem committed:
Make CI happy

yochem committed:
Generate options doc with right text

yochem commented:
asan seems to be fixed, but I can't reproduce the Windows error as I don't have a Windows machine on hand. It looks like `vim.uv.fs_mkdir(testdir .. '/xdg_data', 444)` doesn't create a non-writable directory on Windows.

yochem committed:
Fix CI

yochem committed:
Last cleanup

yochem committed:
Fix test on Windows

justinmk commented on test/functional/core/spellfile_spec.lua:
```diff
@@ -117,4 +117,26 @@ describe('spellfile', function()
+
+    it("is not set if stdpath('data') is not writable", function()
+      n.command('set spell')
+      vim.uv.fs_mkdir(testdir .. '/xdg_data', 444)
```
I think chmod 444 isn't possible on Windows. Instead, to make this test work on windows, try overwriting `.../xdg_data` with a file instead of a directory. That should hopefully "fail" correctly.

If that doesn't work, then use `t.is_os('win')` to `return` from this test (don't use `skip()` because it's not possible to ever "fix" this).

yochem committed:
Fix Windows path separator in test

yochem commented:
I still can't get the Windows test to work. I tried the following:

```
eq(
  vim.fs.normalize(fn.stdpath('data') .. '/spell/en.utf-8.add'),
  vim.fs.normalize(api.nvim_get_option_value('spellfile', {}))
)
```

But that still gives the following error in the test runner:

```
(string) 'Xtest-functional-spell-spellfile.d\xdg_data\nvim-data/spell/en.utf-8.add'
Expected:
(string) 'Xtest-functional-spell-spellfile.d/xdg_data/nvim/spell/en.utf-8.add'
```

Eventhough the docs on `vim.fs.normalize` mention:

> On Windows, backslash (\) characters are converted to forward slashes (/).

justinmk commented:
> Eventhough the docs on `vim.fs.normalize` mention:
> 
> > On Windows, backslash (\) characters are converted to forward slashes (/).

That does seem unexpected, and may be a bug, though we certainly have tests for it...

Note also that `nvim-data/` is the special stdpath('data') path on Windows.

yochem commented:
> Note also that nvim-data/ is the special stdpath('data') path on Windows.

That difference also seems quite weird. C code for getting the data path uses `get_xdg_home(kXDGDataHome);`. Lua code uses `n.fn.stdpath('data')`. Is there a difference between the two?

For the slashes, I'll try testutil's `fix_slashes` which I just accidentally found.

yochem commented:
The Windows tests pass now. There are two (new) failing checks on macos and ubuntu, but I don't think they have anything to do with the changes introduced in this PR.

zeertzjq commented on runtime/doc/vim_diff.txt:
```diff
@@ -79,6 +79,7 @@ Defaults                                            *defaults* *nvim-defaults*
 - 'showcmd' is enabled
 - 'sidescroll' defaults to 1
 - 'smarttab' is enabled
+- 'spellfile' defaults to `stdpath("data").."/spell/"
```
```suggestion
- 'spellfile' defaults to `stdpath("data").."/spell/"`
```

yochem committed:
Update vim_diff.txt

Co-authored-by: zeertzjq <zeertzjq@outlook.com>
